---
title: "*hsrecombi*: workflow to create data for the shiny app CLARITY"
author: "D. Wittenburg, N. Melzer"
date: "(December 17, 2025)"
output: 
 github_document
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = 400
)
```

### Objective
Create a genetic map and interactively visualise the outcome

### Source
The R package CLARITY is available at [https://github.com/nmelzer/CLARITY](https://github.com/nmelzer/CLARITY) 

An online version of CLARITY is available at [https://nmelzer.shinyapps.io/clarity/](https://nmelzer.shinyapps.io/clarity/)


### Instructions
1. The pipeline for creating a male genetic map from genotypes of paternal half-siblings with the 
   hsrecombi approach ([Hampel et al. 2018](https://doi.org/10.3389/fgene.2018.00186)) is available in subfolder makefile\_for\_\<real/sim\>\_data. 
   The pipeline requires make version >= 4.3; this version allows for grouped targeting.
   Ensure that required R packages have been installed ([hsrecombi](https://doi.org/10.32614/CRAN.package.hsrecombi) >=1.0.0, ggplot2; optional: 
   AlphaSimR), then call "make all". For each chromosome i, genetic and physical positions are 
   stored as a list in geneticpositions\_chr\<i\>.Rdata. Further interim results are available.
2. Additionally, sex-specific and average genetic maps employing all genotypes can be estimated 
   with the [LINKPHASE3](https://github.com/tdruet/LINKPHASE3) approach ([Druet & George 2015](https://doi.org/10.1093/bioinformatics/btu859)). 
   To this end, run "run_linkphase.R" in the
   same working directory as "make all" (requires R packages magrittr, hsrecombi, dplyr, data.table).
3. To prepare input data for the R-Shiny app CLARITY version >= 3.0.0 in the required format, run 
   "prepare_data_rshiny.R" in the same working directory as "make all" (requires R packages 
   magrittr, hsrecombi, dplyr). The output Rdata will then be content of folder 
   CLARITY/inst/extdata. Note that, unless working with nchr = 29 bovine autosomes, modifications
   to the app become necessary.


### Citation

[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.11543149.svg)](https://doi.org/10.5281/zenodo.11543149)